### ISSUE TITLE
read.ctd.sbe is very slow for large (>10 MB) cnv files

### ISSUE THREAD

**Issue by richardsc** [dankelley/oce#915]
Investigating where it's getting hung up. Will update more ...

**Comment by richardsc** [comment]
Quick results from doing a profile on a "small" (4.9MB) file:
```r
> Rprof()
> d <- read.ctd(paste0(sbedir, sbefiles[i]))
> Rprof(NULL)
> summaryRprof()
$by.self
                 self.time self.pct total.time total.pct
"textConnection"     10.86    92.03      10.86     92.03
"readLines"           0.58     4.92       0.58      4.92
"scan"                0.14     1.19       0.14      1.19
"read.ctd.sbe"        0.12     1.02      11.80    100.00
".External2"          0.06     0.51       0.06      0.51
"<"                   0.02     0.17       0.02      0.17
"regexpr"             0.02     0.17       0.02      0.17

$by.total
                 total.time total.pct self.time self.pct
"read.ctd.sbe"        11.80    100.00      0.12     1.02
"read.ctd"            11.80    100.00      0.00     0.00
"as.list"             11.06     93.73      0.00     0.00
"read.table"          11.06     93.73      0.00     0.00
"textConnection"      10.86     92.03     10.86    92.03
"readLines"            0.58      4.92      0.58     4.92
"scan"                 0.14      1.19      0.14     1.19
".External2"           0.06      0.51      0.06     0.51
"type.convert"         0.06      0.51      0.00     0.00
"<"                    0.02      0.17      0.02     0.17
"regexpr"              0.02      0.17      0.02     0.17

$sample.interval
[1] 0.02

$sampling.time
[1] 11.8
```

**Comment by richardsc** [comment]
Looks like most of the time is being taken up by the `read.table/as.list` part, e.g. Line [3761](https://github.com/dankelley/oce/blob/develop/R/ctd.R#L3761):
```r
        data <- as.list(read.table(text=lines[seq.int(iline, length(lines))],
                                   header=FALSE, col.names=col.names.inferred))
```

**Comment by richardsc** [comment]
I'm not sure, but I think maybe the problem isn't actually `read.table` or `as.list`, but actually the `textConnection` call that is done when using `read.table(text=...)`.

**Comment by dankelley** [comment]
I never really know whether the act of instrumenting, ipso factor, changes the results. So I wonder if you can try a test, along the lines of
```R
system.time(x<-read.table(file, skip=N))
```
where you pick N after visual inspection of the file. In addition, I'd like to know the ratio of the above time to the number of data read.

Once we have that, maybe the same but with `as.list()` on the table.

These tests will help us to see a lower limit on time, and if it is too long, then we can figure out something better than `read.table`. (There are tricks to read.table, too, e.g. stringsAsFactors and so forth.)

**Comment by richardsc** [comment]
Sure, I'll play around. A quick google turned up this SO page which might be relevant:

http://stackoverflow.com/questions/1727772/quickly-reading-very-large-tables-as-dataframes-in-r

**Comment by dankelley** [comment]
I recall trying `fread` and `read_table` for some other things, and they were problematic. But I think the BIO people use one of those in their (closed-source) ODF package.

**Comment by richardsc** [comment]
See PR #916. I haven't instrumented all the tests you suggest, but the small change I made to use `pushBack` before calling `read.table` on the file directly makes a marked improvement.

I suspect this could be optimized even further -- e.g. by not drinking the entire file with `readLines` but only the header portion, so I'll leave this open for now.

**Comment by richardsc** [comment]
Here is the difference for a 4.9 MB file:

Using `read.table(text=...)`:
```r
   user  system elapsed 
 11.242   0.052  11.331 
```

Using `read.table(file, ...)` after `pushBack(lines, file)`:
```r
   user  system elapsed 
  0.868   0.012   0.880 
```
Which corresponds to a factor of 12 speedup.

**Comment by dankelley** [comment]
I wonder if we should also only read a subset of the lines. How long does it take to `readLines` this 4.9MB file?

**Comment by richardsc** [comment]
0.58 s. See the profile from above. I'll email the file so you can play. Also see the oce-issues/915.

**Comment by dankelley** [comment]
So for a 25MB file I guess we have 5s to get the header. That seems way too long to me.  I can have a look at smarter options for the header part. One way could be to try reading say 2000 lines and if the header ends there, fine (probably very fast) and if not, try reading the full file (rather than trying 4000, then 8000, etc ... that seems silly)

**Comment by richardsc** [comment]
I think that's a good idea.

**Comment by dankelley** [comment]
On my home machine, I'm finding that `readLines` takes about 0.1s per MB. For a 30MB file, then, we burn up 3s in finding the header.  But `readLines` with n=10e3 only takes 0.1s, and so I think there is justification for me to code in something to examine just some fraction of the file, to begin with. The simplest would be to read `n` lines and if there is no header, then just reread the whole file ... this rereading scheme could take 3s for a big file, but the more likely scenario (if `n` is 10e3, say) is to save basically 3s.

I'll code this and push ... but *only* if @richardsc feels he is done editing `R/ctd.R` for a while, since I think we would be working on the same lines of code, otherwise.

**Comment by dankelley** [comment]
I've put some comments also on the pull request https://github.com/dankelley/oce/pull/916#issuecomment-207069008 to the effect that this seems ready for closing now. I think you've done the right fix for this, Clark.

**Comment by richardsc** [comment]
I think this is fixed now, after PR #916 and the associated discussion. Closing.
